<a href="https://colab.research.google.com/github/dipenthakercalypso/Image-and-Text-Labelling-and-Classification-Issues-Resolution/blob/main/MLP_And_Fully_Connected.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Task
Generate a python program to read grayscale image data from MNIST and create a multi layer perceptron with 3 hidden layers, 1 input flattened image layer and 1 output layer.

## Load and preprocess the mnist dataset

### Subtask:
Load the MNIST dataset and preprocess the image data (e.g., normalize pixel values, flatten images).


**Reasoning**:
Load the MNIST dataset, normalize the pixel values, and flatten the images as per the instructions.



In [3]:
from tensorflow.keras.datasets import mnist
import numpy as np

# Load the MNIST dataset
(x_train, y_train), (x_test, y_test) = mnist.load_data()

# Normalize the pixel values
x_train = x_train.astype('float32') / 255.0
x_test = x_test.astype('float32') / 255.0

# Flatten the images
x_train = x_train.reshape((x_train.shape[0], -1))
x_test = x_test.reshape((x_test.shape[0], -1))

print("Training data shape:", x_train.shape)
print("Testing data shape:", x_test.shape)

11490434/11490434 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step
Training data shape: (60000, 784)
Testing data shape: (10000, 784)


## Define the multi-layer perceptron model

### Subtask:
Create a multi-layer perceptron model with the specified architecture (input layer, 3 hidden layers, output layer).


**Reasoning**:
Create a multi-layer perceptron model with the specified architecture.



In [4]:
from tensorflow.keras.layers import Dense, Input
from tensorflow.keras import Model

# Define the input layer
input_layer = Input(shape=(784,))

# Add the hidden layers
hidden_layer_1 = Dense(256, activation='relu')(input_layer)
hidden_layer_2 = Dense(128, activation='relu')(hidden_layer_1)
hidden_layer_3 = Dense(64, activation='relu')(hidden_layer_2)

# Add the output layer
output_layer = Dense(10, activation='softmax')(hidden_layer_3)

# Create the model
model = Model(inputs=input_layer, outputs=output_layer)

# Print the model summary
model.summary()

Model: "functional"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ input_layer (InputLayer)        │ (None, 784)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense (Dense)                   │ (None, 256)            │       200,960 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 128)            │        32,896 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_2 (Dense)                 │ (None, 64)             │         8,256 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_3 (Dense)                 │ (None, 10)             │           650 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 242,762 (948.29 KB)

 Trainable params: 242,762 (948.29 KB)

 Non-trainable params: 0 (0.00 B)

## Compile and train the model

### Subtask:
Compile the model (define optimizer, loss function, and metrics) and train it using the preprocessed MNIST data.


**Reasoning**:
Compile and train the model using the specified optimizer, loss function, metrics, and training parameters.



In [5]:
# Compile the model
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

# Train the model
history = model.fit(x_train, y_train,
                    epochs=10,
                    batch_size=64,
                    validation_split=0.2)

Epoch 1/10
750/750 ━━━━━━━━━━━━━━━━━━━━ 6s 7ms/step - accuracy: 0.8636 - loss: 0.4815 - val_accuracy: 0.9560 - val_loss: 0.1397
Epoch 2/10
750/750 ━━━━━━━━━━━━━━━━━━━━ 5s 6ms/step - accuracy: 0.9679 - loss: 0.1069 - val_accuracy: 0.9678 - val_loss: 0.1084
Epoch 3/10
750/750 ━━━━━━━━━━━━━━━━━━━━ 4s 6ms/step - accuracy: 0.9783 - loss: 0.0665 - val_accuracy: 0.9683 - val_loss: 0.1005
Epoch 4/10
750/750 ━━━━━━━━━━━━━━━━━━━━ 6s 7ms/step - accuracy: 0.9840 - loss: 0.0523 - val_accuracy: 0.9715 - val_loss: 0.0953
Epoch 5/10
750/750 ━━━━━━━━━━━━━━━━━━━━ 10s 7ms/step - accuracy: 0.9891 - loss: 0.0363 - val_accuracy: 0.9765 - val_loss: 0.0856
Epoch 6/10
750/750 ━━━━━━━━━━━━━━━━━━━━ 5s 6ms/step - accuracy: 0.9904 - loss: 0.0294 - val_accuracy: 0.9741 - val_loss: 0.0977
Epoch 7/10
750/750 ━━━━━━━━━━━━━━━━━━━━ 4s 6ms/step - accuracy: 0.9922 - loss: 0.0237 - val_accuracy: 0.9693 - val_loss: 0.1197
Epoch 8/10
750/750 ━━━━━━━━━━━━━━━━━━━━ 6s 7ms/step - accuracy: 0.9938 - loss: 0.0211 - val_accuracy: 0

## Evaluate the model

### Subtask:
Evaluate the trained model on the test set to assess its performance.


**Reasoning**:
Evaluate the trained model on the test set and print the results.



In [6]:
# Evaluate the model on the test data
loss, accuracy = model.evaluate(x_test, y_test, verbose=0)

# Print the evaluation results
print(f'Test loss: {loss:.4f}')
print(f'Test accuracy: {accuracy:.4f}')

Test loss: 0.0781
Test accuracy: 0.9804


## Summary:

### Data Analysis Key Findings

*   The MNIST dataset was successfully loaded, normalized, and flattened, resulting in training data with a shape of (60000, 784) and testing data with a shape of (10000, 784).
*   A multi-layer perceptron model was constructed with an input layer (784 units), three hidden layers (256, 128, and 64 units with ReLU activation), and an output layer (10 units with softmax activation).
*   The model was compiled using the 'adam' optimizer and 'sparse\_categorical\_crossentropy' loss function, and trained for 10 epochs with a batch size of 64 and a 20% validation split.
*   During training, the validation accuracy reached approximately 98.03% and the validation loss was about 0.0930 after 10 epochs.
*   Upon evaluation on the test set, the model achieved a test loss of 0.0781 and a test accuracy of 0.9804.

### Insights or Next Steps

*   The model shows strong performance on the MNIST digit classification task, achieving high accuracy on both validation and test sets.
*   Further improvements could potentially be explored by experimenting with different network architectures, hyperparameter tuning, or regularization techniques to potentially reduce the test loss further.
